### PR TITLE
Hungarian TN ✅

### PR DESCRIPTION
**Pull request by jimregan**
# What does this PR do ?

Adds TN for Hungarian

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Taggers:
  - [X] Cardinals
  - [X] Ordinals
  - [x] Date
  - [x] Time
  - [X] Electronic
  - [x] Measure
  - [x] Decimals
  - [X] Fraction
  - [x] Telephone
  - [X] Whitelist
  - [x] Money
 
- Verbalisers:
  - [X] Cardinals
  - [x] Ordinals
  - [x] Date
  - [x] Time
  - [X] Electronic
  - [x] Measure
  - [X] Decimals
  - [X] Fraction
  - [x] Telephone
  - [x] Whitelist
  - [x] Money

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [X] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [X] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Comment by jimregan** [comment]
> @jimregan , thanks for the PR. Do you know when the PR is ready for review? I see you wanted to add some more classes like time, measure and money. Once you are ready I will ask a Hungarian native speaker review the `data` and `test case ` files

Swedish TN was a work project, so I had more time to work on it. I ought to have some time this evening and this weekend -- I'd like to get it finished quite soon, but I can't make any guarantees. If you'd like, I'll see how far I get by Monday, drop out anything that's left unfinished, and test what's left. There's not much left to do, IIRC, but I wanted to get the more informal way of reading times done for future use (ITN/audio-based TN), and to copy for Swedish ITN (which does something very similar).

Hungarian is standardised quite extensively, so for most of the modules I never really had too many questions, but I do have access to native speakers here. One more can't hurt, though.

There are a couple of decisions that are kind of debatable: for one, in the case of money, I set it to read the amount as a decimal if there is a case ending at the end of the currency chunk, rather than try to transfer it to the currency minor word. It wouldn't be too hard to do that, but the colleague I asked read it as a decimal, so that's what I went with.

**Comment by jimregan** [comment]
'Ready for review' is not 100% accurate, as more tests are needed, and there's an issue with things not quite working the same in sparrowhawk, but insofar as being more-or-less ready to have an extra pair of eyeballs, it's close enough

**Comment by yzhang123** [comment]
> 'Ready for review' is not 100% accurate, as more tests are needed, and there's an issue with things not quite working the same in sparrowhawk, but insofar as being more-or-less ready to have an extra pair of eyeballs, it's close enough

could you elaborate what does not work with sparrowhawk?

**Comment by jimregan** [comment]
> > 'Ready for review' is not 100% accurate, as more tests are needed, and there's an issue with things not quite working the same in sparrowhawk, but insofar as being more-or-less ready to have an extra pair of eyeballs, it's close enough
> 
> could you elaborate what does not work with sparrowhawk?

```
ASSERT:100 000 000 000 000 000 000 002 expected:<száztrilliárd-kettő> but was:<száztrilliárd------kettő>
```
I have a piece that normally replaces repeated hyphens with a single hyphen, but this isn't happening with sparrowhawk.

File: nemo_text_processing/text_normalization/hu/data/electronic/symbols.tsv [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+#	kettőskereszt
+$	dollárjel
+%	százalékjel
+&	és
+'	aposztróf
+*	csillagjel
++	pluszjel
+/	perjel
+\	fordított perjel
+=	egyenlő
```
Comment (Laszlo-Weber) [inline nemo_text_processing/text_normalization/hu/data/electronic/symbols.tsv:14]: It is not incorrect just not consistent with the other. I would use: egyenlőségjel

File: nemo_text_processing/text_normalization/hu/data/electronic/symbols.tsv [hunk 2]
```diff
@@ -0,0 +1,23 @@
+.	pont
+-	kötőjel
+_	aláhúzásjel
+!	felkiáltójel
+#	kettőskereszt
+$	dollárjel
+%	százalékjel
+&	és
```
Comment (Laszlo-Weber) [inline nemo_text_processing/text_normalization/hu/data/electronic/symbols.tsv:8]: It is not incorrect just not consistent with the others. I would use: ésjel

File: nemo_text_processing/text_normalization/hu/data/measures/measurements.tsv [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+m²	négyzetméter
+m2	négyzetméter
+m³	köbméter
+m3	köbméter
+mph	mérföld per óra
+mi/h	mérföld per óra
+ml	milliliter
+mm	milliméter
+ms	milliszekundum
+%	százalék
```
Comment (Laszlo-Weber) [inline nemo_text_processing/text_normalization/hu/data/measures/measurements.tsv:51]: The line is correct but why % is a measurement unit?

**Comment by yzhang123** [comment]
@jimregan  we have reviewed the data, and it looks good. Please rerequest review once this PR is ready to be merged. thanks!

**Comment by jimregan** [comment]
> @jimregan we have reviewed the data, and it looks good. Please rerequest review once this PR is ready to be merged. thanks!

There are a couple of small things: the kestrel difference I mentioned earlier, and one small issue with serialisation. I think there might be an easy with the second, but the first has me baffled. If I can find a moment, I'll see if I can do something about it, but I don't expect to have time until the start of march.

**Comment by yzhang123** [comment]
> > @jimregan we have reviewed the data, and it looks good. Please rerequest review once this PR is ready to be merged. thanks!
> 
> There are a couple of small things: the kestrel difference I mentioned earlier, and one small issue with serialisation. I think there might be an easy with the second, but the first has me baffled. If I can find a moment, I'll see if I can do something about it, but I don't expect to have time until the start of march.

cool, thanks! no rush.

**Comment by jimregan** [comment]
Still the same issue with sparrowhawk:
> ASSERT:100 000 000 000 000 000 000 002 expected:<száztrilliárd-kettő> but was:<száztrilliárd------kettő>

**Comment by yzhang123** [comment]
> Still the same issue with sparrowhawk:
> 
> > ASSERT:100 000 000 000 000 000 000 002 expected:<száztrilliárd-kettő> but was:<száztrilliárd------kettő>

@anand-nv could you please help?

**Comment by jimregan** [comment]
> > Still the same issue with sparrowhawk:
> > > ASSERT:100 000 000 000 000 000 000 002 expected:<száztrilliárd-kettő> but was:<száztrilliárd------kettő>
> 
> @anand-nv could you please help?

There's more going on here: the tests pass when run individually, but not when the whole directory is run. 

E.g., `pytest --cpu tests/nemo_text_processing/hu`
gives:
> E       AssertionError: assert 'nullatrilliá...armincegyedik' == 'harmincegyedik'
> E         - harmincegyedik
> E         + nullatrilliárd-harmincegyedik

but with `pytest --cpu tests.nemo_text_processing.hu.test_ordinal` it passes.

**Comment by jimregan** [comment]
> > Still the same issue with sparrowhawk:
> > > ASSERT:100 000 000 000 000 000 000 002 expected:<száztrilliárd-kettő> but was:<száztrilliárd------kettő>
> 
> @anand-nv could you please help?

This one's fixed, I just have to fix the ones I managed to introduce along the way :)

**Comment by jimregan** [comment]
```
self = <tests.nemo_text_processing.hu.test_money.TestMoney object at 0x7f48cd070520>, test_input = '2,32 GBP-t'
expected = 'kettő egész harminckettő század angol fontot'

    @parameterized.expand(parse_test_case_file('hu/data_text_normalization/test_cases_money.txt'))
    @pytest.mark.run_only_on('CPU')
    @pytest.mark.unit
    def test_norm(self, test_input, expected):
        pred = self.normalizer_hu.normalize(test_input, verbose=False)
>       assert pred == expected
E       AssertionError: assert 'nullatrillió... angol fontot' == 'kettő egész ... angol fontot'
E         - kettő egész harminckettő század angol fontot
E         + nullatrillió-kettő egész nullatrillió-harminckettő század angol fontot
E         ? +++++++++++++            +++++++++++++
```

The weird thing seems to depend on casing:

```
>>> from nemo_text_processing.text_normalization.normalize import Normalizer
>>> lc_normalizer = Normalizer(input_case='lower_cased', lang='hu', overwrite_cache=False)
>>> normalizer = Normalizer(input_case='cased', lang='hu', overwrite_cache=False)
>>> normalizer.normalize("2,32 GBP-t")
'nullatrillió-kettő egész nullatrillió-harminckettő század angol fontot'
>>> lc_normalizer.normalize("2,32 GBP-t")
'kettő egész harminckettő század angol fontot'
```

**Comment by jimregan** [comment]
> The weird thing seems to depend on casing:
> 

Rather, the first instance is fine, subsequent instances have something weird happening:

```
Python 3.8.0 (default, Nov  6 2019, 21:49:08) 
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from nemo_text_processing.text_normalization.normalize import Normalizer
>>> tmp = []
>>> for _ in range(0, 6):
...     tmp.append(Normalizer(input_case='lower_cased', lang='hu'))
... 
>>> for norm in tmp:
...     norm.normalize("2,32 GBP-t")
... 
'kettő egész harminckettő század angol fontot'
'nullatrillió-kettő egész nullatrillió-harminckettő század angol fontot'
'nullatrillió-kettő egész nullatrillió-harminckettő század angol fontot'
'nullatrillió-kettő egész nullatrillió-harminckettő század angol fontot'
'nullatrillió-kettő egész nullatrillió-harminckettő század angol fontot'
'nullatrillió-kettő egész nullatrillió-harminckettő század angol fontot'
```

@yzhang123 any ideas why this might be happening?

**Comment by jimregan** [comment]
> > The weird thing seems to depend on casing:
> 
> Rather, the first instance is fine, subsequent instances have something weird happening:
> 

cardinal is being built differently second time around:

```
>>> tmp[0].tagger.cardinal.fst.num_states()
2253
>>> tmp[1].tagger.cardinal.fst.num_states()
3335
```
Anything that doesn't depend on cardinal is identical.

**Comment by jimregan** [comment]
Pytest:
```
=============================================== 199 passed, 3 warnings in 283.31s (0:04:43) ===============================================
```

Sparrowhawk:
```
Step 15/15 : RUN echo "DONE"
 ---> Using cache
 ---> 81110575ab3c
Successfully built 81110575ab3c
Successfully tagged sparrowhawk:latest
-v /home/joregan/NeMo-text-processing/tools/text_processing_deployment/docker/../hu/classify:/workspace/sparrowhawk/documentation/grammars/en_toy/classify -v /home/joregan/NeMo-text-processing/tools/text_processing_deployment/docker/../hu/verbalize:/workspace/sparrowhawk/documentation/grammars/en_toy/verbalize
testTNCardinal
testTNDate
testTNDecimal
testTNElectronic
testTNFraction
testTNMoney
testTNOrdinal
testTNTelephone
testTNTime
testTNMeasure
testTNWhitelist
testTNWord

Ran 12 tests.

OK
```

@yzhang123 this is done

**Comment by jimregan** [comment]
@yzhang123 @ekmb 👋

**Comment by yzhang123** [comment]
thanks!

**Comment by jimregan** [comment]
@Laszlo-Weber sorry about that, Chrome froze and apparently the mouse cursor landed there in the interim

**Review by yzhang123 — APPROVED** [review]

**Comment by yzhang123** [comment]
@jimregan how did you fix the sparrowhawk issues?

**Comment by jimregan** [comment]
> @jimregan how did you fix the sparrowhawk issues?

I think it was this: https://github.com/NVIDIA/NeMo-text-processing/pull/9/commits/71cbfe7004d3f990a15e2b5ea43bc22e1a565ae1 -- I moved the '|=' with '000' and then everything worked, but I'd changed everything I could think of by that stage, so I can't say for certain it was that

**Pull request merged**